File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int max=Integer.MIN_VALUE;
        int p=1;
        for(int i=0;i<nums.length;i++){
            p=1;
            for(int j=i;j<nums.length;j++){
                p*=nums[j];
                max=Math.max(max,p);
            }
        }
        return max;
    }}
    // static int p(int i,int j,int[] nums){
    //     int[] pi=new int[nums.length];
    //     pi[i]=nums[i];
    //     for(int k=i+1;k<=j;k++){
    //         pi[k]=pi[k-1]*nums[k];
    //     }
    //     return pi[j];
    // }

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # pmax=smax=pc=sc=nums[0]
        pc=sc=1
        pmax=smax=-10**100
        # if pc==0:pc=1
        # if sc==0:sc=1
        for i in range(len(nums)):
            pc*=nums[i]
            sc*=nums[len(nums)-i-1]
            pmax=max(pmax,pc)
            smax=max(smax,sc)
            if pc==0:
                pc=1
            if sc==0:
                sc=1
        return max(pmax,smax)